### PR TITLE
Coverage in TF import

### PR DESCRIPTION
**Pull request by AlexDBlack**

### CODE REVIEW

File: tests/test_ops.py [hunk 1]
```diff
@@ -1239,28 +1240,36 @@ def test_mathtransform():
...
         # {"opName": "fake_quant_with_min_max_vars_per_channel", "outName": "fake_quant/min_max_args_per_channel/rank2_8bit_narrow", "varShapes":[[3, 5], [5], [5]], "varTypes":["float32", "float32", "float32"], "varInit":["uniform", "uniform_m1_0", "uniform"], "num_bits":8, "narrow_range":True},
         # {"opName": "fake_quant_with_min_max_vars_per_channel", "outName": "fake_quant/min_max_args_per_channel/rank4_6bit", "varShapes":[[3, 2, 2, 5], [5], [5]], "varTypes":["float32", "float32", "float32"], "varInit":["uniform", "uniform_m1_0", "uniform"], "num_bits":6, "narrow_range":False},
 
+        #{"opName": "adjust_contrast", "outName": "adjust_contrast", "varShapes": [[256,256,3], ()],"varTypes": ["float32", "float32"], "varInit": ["uniform", "uniform"]},
+        #{"opName": "adjust_contrast_v2", "outName": "adjust_contrast_v2", "varShapes": [[256,256,3], ()],"varTypes": ["float32", "float32"], "varInit": ["uniform", "uniform"]},
```
Comment (AlexDBlack) [inline tests/test_ops.py]: Use smaller arrays. Keep in mind we're storing this stuff in text format. Something like `[8,8,3]` would be much smaller to store (and easier to debug if it fails) but should still check correctness.

Also add more cases. If you read the op: it supports rank 3+ arrays.
https://www.tensorflow.org/api_docs/python/tf/image/adjust_contrast

If I was doing this I would do the following shapes: [8,8,3], [5,8,3], [2, 5, 5, 3] - rank 3, rank 3 (non-square), and rank 4.
If the op supported multiple datatypes (it doesn't) we'd need tests for that too.

File: tfoptests/ops.py [hunk 2]
```diff
@@ -1227,8 +1227,42 @@ def execute_fake_quant_with_min_max_vars_per_channel(self):
     def execute_check_numerics(self):
         return [tf.debugging.check_numerics(tensor=self.vars[0], message=self.op["message"])]
 
+    def execute_adjust_contrast(self):
+        return [tf.compat.v1.image.adjust_contrast(self.vars[0],self.vars[1])]
+
+    def execute_adjust_contrast_v2(self):
+        return [tf.compat.v2.image.adjust_contrast(self.vars[0],self.vars[1])]
```
Comment (AlexDBlack) [inline tfoptests/ops.py:1237]: `tf.compat.v1.image.adjust_contrast` and `tf.compat.v2.image.adjust_contrast` are the exact same op AFAIK, not v1 and v2 of the op or anything. So one of these is redundant.

File: tfoptests/ops.py [hunk 3]
```diff
@@ -1227,8 +1227,42 @@ def execute_fake_quant_with_min_max_vars_per_channel(self):
...
+
     def execute_strings_split(self):
         print("strings.split input: ", self.vars[0])
         out = tf.strings.split(self.vars[0], sep=self.op["split"])
         print("strings.split output: ", out)
         return [out]
+
+    def execute_bitcast(self):
+        return [tf.bitcast(self.vars[0], tf.int32)]
```
Comment (AlexDBlack) [inline tfoptests/ops.py]: Avoid hardcoding arguments/types like this. We can't get proper coverage or testing.
The op itself supports a ton of different output types: https://www.tensorflow.org/versions/r1.15/api_docs/python/tf/bitcast?hl=en

For the extra args, do it like this:
https://github.com/KonduitAI/ImportTests/blob/master/tests/test_ops.py#L988-L989
https://github.com/KonduitAI/ImportTests/blob/master/tests/test_ops.py#L1390

Note `num=self.op["num"]` in the use and ` "num":4,` in the definition. It's just a dictionary here, you can put whatever you want in there and use `self.op[<key>]` in the op definition.

File: tests/test_ops.py [hunk 4]
```diff
@@ -1734,7 +1743,12 @@ def test_mathtransform():
         # {"opName": "reduce_mean", "outName": "ragged/reduce_mean/2d_a1", "varShapes":[[]], "varTypes":["float32"], "varInit":["ragged2d"], "axis":1, "keepdims":False},
         # {"opName": "sqrt", "outName": "ragged/sqrt/2d", "varShapes":[[]], "varTypes":["float32"], "varInit":["ragged2d"]},
 
-        {"opName": "strings_split", "outName": "ragged/sqrt/2d", "varShapes":[[]], "varTypes":["string"], "varInit":["string2"], "split":" "},
+        # {"opName": "strings_split", "outName": "ragged/sqrt/2d", "varShapes":[[]], "varTypes":["string"], "varInit":["string2"], "split":" "},
+        # {"opName": "bitcast", "outName": "bitcast", "varShapes":[[1,10]], "varTypes":["float32"], "varInit":["uniform", "uniform"], "dtype":tf.float32},
```
Comment (AlexDBlack) [inline tests/test_ops.py]: As noted below: Bitcast op needs considerably better coverage.
The op supports:
* a large number of input types
* a large number of output types
* Different ranks

For proper coverage, we want some combination of those.


At a minimum, I'd want a collection of 5-8 test cases along the lines of:
* float32 -> int32, rank 0 scalar
* float64 -> int32, rank 1
* int64 -> int8, rank 2
* uint16 -> bfloat16, rank 1

File: tests/test_ops.py [hunk 5]
```diff
@@ -1734,7 +1743,12 @@ def test_mathtransform():
...
+        # {"opName": "bitwise_or", "outName": "bitwise_or", "varShapes":[[1,10], [1,10]], "varTypes":["int64", "int64"], "varInit":["uniform_int10", "uniform_int10"]},
+        # {"opName": "bitwise_xor", "outName": "bitwise_xor", "varShapes":[[1,10], [1,10]], "varTypes":["int64", "int64"], "varInit":["uniform_int10", "uniform_int10"]},
```
Comment (AlexDBlack) [inline tests/test_ops.py]: Bitwise and/or: again, proper coverage.
The ops supports int8-64 and uint8-uint64.
We don't necessarily need test cases for every single datatype there, but we definitely need more than one.
When generating different cases, also vary the shapes. For example an int8 scalar, uint16 rank 2, etc etc

File: tests/test_ops.py [hunk 6]
```diff
@@ -1734,7 +1743,12 @@ def test_mathtransform():
...
+        # {"opName": "is_non_decreasing", "outName": "is_non_decreasing", "varShapes":[[1,10]], "varTypes":["float32"], "varInit":["uniform"]},
+        #  {"opName": "is_strictly_increasing", "outName": "is_strictly_increasing", "varShapes":[[1,10]], "varTypes":["float32"], "varInit":["uniform"]},
```
Comment (AlexDBlack) [inline tests/test_ops.py]: Again, improved coverage. Rank 0, 1, 2 + different datatypes (int8, int32, bfloat16, float32, float64)
Also be careful with the initialization here. Uniform initialization on [1,10] array will almost always be non-decreasing (not increasing) in practice -> we're never testing one side of the condition.

File: tfoptests/ops.py [hunk 7]
```diff
@@ -1227,8 +1227,42 @@ def execute_fake_quant_with_min_max_vars_per_channel(self):
...
+        return [tf.nn.dropout(x = self.vars[0], rate = 0.25)]
+
+    def execute_is_strictly_increasing(self):
+        return [tf.math.is_strictly_increasing(x = self.vars[0])]
+
+    def execute_leaky_relu(self):
+        return tf.nn.leaky_relu(features = self.vars[0], alpha=self.op["alpha"])
+
+    def execute_Conv2DTranspose(self):
+        return [tf.keras.layers.Conv2DTranspose(kernel_size=self.op["kernel_size"], filters=[self.op["filters"]])]
```
Comment (AlexDBlack) [inline tfoptests/ops.py]: Don't use tf.keras.

File: tests/test_ops.py (hunk 1 above)
Comment (alexanderst) [reply tests/test_ops.py]: Implemented.

File: tfoptests/ops.py (hunk 2 above)
Comment (alexanderst) [reply tfoptests/ops.py:1237]: Fixed.

File: tfoptests/ops.py (hunk 3 above)
Comment (alexanderst) [reply tfoptests/ops.py]: Fixed.

File: tests/test_ops.py (hunk 4 above)
Comment (alexanderst) [reply tests/test_ops.py]: Fixed.

File: tests/test_ops.py (hunk 5 above)
Comment (alexanderst) [reply tests/test_ops.py]: Fixed.

File: tfoptests/ops.py (hunk 7 above)
Comment (alexanderst) [reply tfoptests/ops.py]: Fixed.

File: tests/test_ops.py (hunk 6 above)
Comment (alexanderst) [reply tests/test_ops.py]: Fixed

File: tests/test_ops.py [hunk 8]
```diff
@@ -1239,28 +1240,44 @@ def test_mathtransform():
...
         # {"opName": "fake_quant_with_min_max_vars_per_channel", "outName": "fake_quant/min_max_args_per_channel/rank4_6bit", "varShapes":[[3, 2, 2, 5], [5], [5]], "varTypes":["float32", "float32", "float32"], "varInit":["uniform", "uniform_m1_0", "uniform"], "num_bits":6, "narrow_range":False},
 
+         #{"opName": "adjust_contrast", "outName": "adjust_contrast/rank3_float32", "varShapes": [[8,8,3], []],"varTypes": ["float32", "float32"], "varInit": ["uniform", "uniform", "two"]},
+         #{"opName": "adjust_contrast", "outName": "adjust_contrast_v/rank3_mixed", "varShapes": [[8,8,3,1], []],"varTypes": ["float32", "float32"], "varInit": ["uniform", "two"]},
+         #{"opName": "adjust_contrast", "outName": "adjust_contrast_v/rank4_mixed","varShapes": [[8, 8, 3, 4], []], "varTypes": ["float32", "float32"], "varInit": ["uniform", "two"]},
+
+        #{"opName": "crop_and_resize", "outName": "crop_and_resize", "varShapes": [[1,2,2,1], [2,4], [2], [2]],"varTypes": ["float32", "float32","int32","int32"], "varInit": ["uniform", "uniform","uniform_int10","uniform_int10"]},
```
Comment (AlexDBlack) [inline tests/test_ops.py]: Need more coverage for:
* Different datatypes
* More shapes
* method (bilinear vs. nearest)
* extrapolation value

https://www.tensorflow.org/versions/r1.15/api_docs/python/tf/image/crop_and_resize

File: tests/test_ops.py [hunk 9]
```diff
@@ -1239,28 +1240,44 @@ def test_mathtransform():
...
+
+
+        # {"opName": "draw_bounding_boxes", "outName": "draw_bounding_boxes", "varShapes": [[2,256,256,1], [2,2,4]],"varTypes": ["float32", "float32"], "varInit": ["uniform", "uniform"]},
```
Comment (AlexDBlack) [inline tests/test_ops.py]: Smaller size. Like [2, 20, 20, 1].
Also test half datatype.

File: tests/test_ops.py [hunk 10]
```diff
@@ -1239,28 +1240,44 @@ def test_mathtransform():
...
+
+        # {"opName": "resize_bilinear", "outName": "resize_bilinear/float32", "varShapes": [[2,256,256,1], [2]],"varTypes": ["float32", "int32"], "varInit": ["uniform", "uniform_int2"]},
+        # {"opName": "resize_bilinear", "outName": "resize_bilinear/float64", "varShapes": [[2, 256, 256, 1], [2]],  "varTypes": ["float64", "int32"], "varInit": ["uniform", "uniform_int2"]},
```
Comment (AlexDBlack) [inline tests/test_ops.py]: Again, smaller arrays.
Also this initialization looks bad: `"uniform_int2"` - that's U(0,2) - which means height/width can only be 0, 1 or 2 out... something like uniform_int10 would be much better, or even add a new initialization.
Also a [x,y,z,3] case would be good (NHWC format - C=1 is grayscale, C=3 is RGB)

2 other attributes not tested:
* align_corners
* half_pixel_centers

https://www.tensorflow.org/versions/r1.15/api_docs/python/tf/image/resize_bilinear

TF docs are bad, but if you check ops.proto, images (first array) datatype can be many different types; including integer types. We need to test more of these.

File: tests/test_ops.py [hunk 11]
```diff
@@ -1239,28 +1240,44 @@ def test_mathtransform():
...
+
+        # {"opName": "resize_nearest_neighbor", "outName": "resize_nearest_neighbor/float32", "varShapes": [[2,256,256,1], [2]],"varTypes": ["float32", "int32"], "varInit": ["uniform", "uniform_int10"]},
+        #  {"opName": "resize_nearest_neighbor", "outName": "resize_nearest_neighbor/float64", "varShapes": [[2,256,256,1], [2]],"varTypes": ["float64", "int32"], "varInit": ["uniform", "uniform_int10"]},
```
Comment (AlexDBlack) [inline tests/test_ops.py]: as per resize_bilinear (except for init which is OK)

File: tests/test_ops.py [hunk 12]
```diff
@@ -1239,28 +1240,44 @@ def test_mathtransform():
...
 
         # {"opName": "check_numerics", "outName": "check_numerics/rank1_float16", "varShapes":[[5]], "varTypes":["float16", "string"], "varInit":["uniform", "string_scalar"], "message":"This is a test string."},
         # {"opName": "check_numerics", "outName": "check_numerics/rank1_float32", "varShapes":[[5]], "varTypes":["float32", "string"], "varInit":["uniform", "string_scalar"], "message":"This is a test string."},
         # {"opName": "check_numerics", "outName": "check_numerics/rank2_float64", "varShapes":[[5]], "varTypes":["float64", "string"], "varInit":["uniform", "string_scalar"], "message":"This is a test string."},
         # {"opName": "check_numerics", "outName": "check_numerics/rank4_bfloat16", "varShapes":[[5]], "varTypes":["bfloat16", "string"], "varInit":["uniform", "string_scalar"], "message":"This is a test string."}
 
+        # {"opName": "dropout", "outName": "dropout", "varShapes":[[1,100]], "varTypes":["float32"], "varInit":["uniform"]}
```
Comment (AlexDBlack) [inline tests/test_ops.py:1286]: One problem with testing dropout: no way to validate results, due to randomness...
And I think this is implemented as multiple ops internally, not one op.
So not a high priority to test...

File: tests/test_ops.py [hunk 13]
```diff
@@ -1395,14 +1412,29 @@ def test_mathtransform():
...
         # {"opName": "zeros_like", "outName": "emptyArrayTests/zeros_like/rank1", "varShapes":[[0]], "varTypes":["float32"], "varInit":["empty"]},
         # {"opName": "zeros_like", "outName": "emptyArrayTests/zeros_like/rank2", "varShapes":[[2,0]], "varTypes":["float64"], "varInit":["empty"]},
+        # {"opName": "zeros_like", "outName": "zeros_like/rank2_float16", "varShapes": [[3, 12]],"varTypes": ["float16"], "varInit": ["uniform"]},
+        #{"opName": "zeros_like", "outName": "zeros_like/rank2_float32", "varShapes": [[3, 12]], "varTypes": ["float32"], "varInit": ["uniform"]},
+        #{"opName": "zeros_like", "outName": "zeros_like/rank2_float64", "varShapes": [[3, 12]], "varTypes": ["float64"], "varInit": ["uniform"]},
+        #{"opName": "zeros_like", "outName": "zeros_like/rank2_int32", "varShapes": [[3, 10]], "varTypes": ["int32"], "varInit": ["uniform_int10"]},
+        #{"opName": "zeros_like", "outName": "zeros_like/rank2_int64", "varShapes": [[3, 10]], "varTypes": ["int64"], "varInit": ["uniform_int10"]},
```
Comment (AlexDBlack) [inline tests/test_ops.py]: What about adding an empty array case? Rank 0 scalar array case? Boolean case?
It also has optional dtype arg, for output type, we'd want at least one test setting that (to different to input type)

File: tests/test_ops.py [hunk 14]
```diff
@@ -1395,14 +1412,29 @@ def test_mathtransform():
...
+        #{"opName": "abs", "outName": "abs/rank2_int32", "varShapes": [[1, 10]], "varTypes": ["int32"], "varInit": ["uniform_int10"], "dtype": tf.int32},
+        #{"opName": "abs", "outName": "abs/rank2_int64", "varShapes": [[1, 10]], "varTypes": ["int64"], "varInit": ["uniform_int10"],"dtype": tf.int64},
+        #{"opName": "abs", "outName": "abs/rank3_float64", "varShapes": [[1, 10, 20]], "varTypes": ["float64"], "varInit": ["uniform"], "dtype": tf.float64},
+        #{"opName": "abs", "outName": "abs/rank3_float32", "varShapes": [[1, 10, 20]], "varTypes": ["float32"], "varInit": ["uniform"], "dtype": tf.float32},
+        #{"opName": "abs", "outName": "abs/rank3_float16", "varShapes": [[1, 10, 20]], "varTypes": ["float16"], "varInit": ["uniform"], "dtype": tf.float16},
+        #{"opName": "abs", "outName": "abs/rank3_int32", "varShapes": [[1, 10, 20]], "varTypes": ["int32"], "varInit": ["uniform_int10"], "dtype": tf.int32},
+        #{"opName": "abs", "outName": "abs/rank3_int64", "varShapes": [[1, 10, 20]], "varTypes": ["int64"], "varInit": ["uniform_int10"], "dtype": tf.int64},
```
Comment (AlexDBlack) [inline tests/test_ops.py]: abs op: this is absolute value operation.
"uniform" and "uniform_intX" initializations are all positive.
Initializations must include some negative and some positive values to properly test the op.

File: tests/test_ops.py [hunk 15]
```diff
@@ -1417,6 +1449,10 @@ def test_mathtransform():
         # TODO: Div just gives "TypeError: unsupported operand type(s) for /: 'list' and 'list'" event through other ops work :/
         # {"opName": "div", "outName": "emptyArrayTests/div/rank1", "varShapes":[[0], [0]], "varTypes":["float32", "float32"], "varInit":["empty","empty"], "dtype":tf.float32},
         # {"opName": "div", "outName": "emptyArrayTests/div/rank2", "varShapes":[[0,2], [0,2]], "varTypes":["float64", "float64"], "varInit":["empty", "empty"], "dtype":tf.float64},
+        #  {"opName": "div", "outName": "div", "varShapes":[[1,10], [1,10]], "varTypes":["float64", "float64"], "varInit":["uniform", "uniform"], "dtype":tf.float64},
+
+        #{"opName": "div_no_nan", "outName": "div_no_nan", "varShapes":[[1,10], [1,10]], "varTypes":["float32", "float32"], "varInit":["uniform","uniform"], "dtype":tf.float32},
+        #{"opName": "div_no_nan", "outName": "div_no_nan", "varShapes": [[1, 10], [1, 10]],"varTypes": ["float64", "float64"], "varInit": ["uniform", "uniform"], "dtype": tf.float64},
```
Comment (AlexDBlack) [inline tests/test_ops.py:1471]: Docs and ops.proto aren't clear - but I would expect this op to support broadcasting.

File: tests/test_ops.py [hunk 16]
```diff
@@ -1529,6 +1565,26 @@ def test_mathtransform():
...
+        # {"opName": "max", "outName": "max/rank3_float32", "varShapes": [[2, 3, 12], [2, 3, 12]], "varTypes": ["float32", "float32"],  "varInit": ["uniform", "uniform"]},
+        # {"opName": "max", "outName": "max/rank3_float64", "varShapes": [[2, 3, 10], [2, 3, 10]],  "varTypes": ["float64", "float64"], "varInit": ["uniform", "uniform"]},
+        # {"opName": "max", "outName": "max/rank3_float64", "varShapes": [[2, 3, 12], [2, 3, 12]], "varTypes": ["float64", "float64"], "varInit": ["uniform", "uniform"]},
+        # {"opName": "max", "outName": "max/rank3_float64", "varShapes": [[2, 3, 12], [2, 3, 12]], "varTypes": ["float64", "float64"], "varInit": ["uniform", "uniform"]},
+        # {"opName": "max", "outName": "max/rank3_half", "varShapes": [[2, 3, 10], [2, 3, 10]],  "varTypes": ["half", "half"], "varInit": ["uniform", "uniform"]},
+        # {"opName": "max", "outName": "max/rank3_half", "varShapes": [[2, 3, 12], [2, 3, 12]], "varTypes": ["half", "half"], "varInit": ["uniform", "uniform"]},
+        # {"opName": "max", "outName": "max/rank3_half", "varShapes": [[2, 3, 12], [2, 3, 12]], "varTypes": ["half", "half"], "varInit": ["uniform", "uniform"]},
```
Comment (AlexDBlack) [inline tests/test_ops.py:1602]: No broadcast cases, we'll want at least some of those. You can just change some of the shapes for the existing test cases here
For example `[[2, 1, 10], [2, 3, 1]]` (to give [2,3,10] output shape), or `[[], [2,12]` for scalar broadcast

> NOTE: math.maximum supports broadcasting.

https://www.tensorflow.org/api_docs/python/tf/math/maximum?hl=sl

File: tests/test_ops.py [hunk 17]
```diff
@@ -1734,9 +1790,241 @@ def test_mathtransform():
...
+
+        #{"opName": "bitcast", "outName": "bitcast/from_float32_to_float64", "varShapes": [[2]], "varTypes": ["float32"], "varInit": ["uniform"], "output": tf.float64},
+        #{"opName": "bitcast", "outName": "bitcast/from_bfloat16_to_float64", "varShapes": [[4]], "varTypes": ["bfloat16"],"varInit": ["uniform"], "output": tf.float64},
+        #{"opName": "bitcast", "outName": "bitcast/from_half_to_float64", "varShapes": [[4]], "varTypes": ["half"], "varInit": ["uniform"], "output": tf.float64},
+        #{"opName": "bitcast", "outName": "bitcast/from_float64_to_float64", "varShapes": [[1]], "varTypes": ["float64"], "varInit": ["uniform"], "output": tf.float64},
+        #{"opName": "bitcast", "outName": "bitcast/from_int32_to_float64", "varShapes": [[2]], "varTypes": ["int32"], "varInit": ["uniform_int2"], "output": tf.float64},
+        #{"opName": "bitcast", "outName": "bitcast/from_int64_to_float64", "varShapes": [[2]], "varTypes": ["int64"], "varInit": ["uniform_int2"], "output": tf.float64},
+        #{"opName": "bitcast", "outName": "bitcast/emptyArrayTest/from_uint32_to_float64", "varShapes": [[0]], "varTypes": ["uint32"], "varInit": ["empty"], "output": tf.float64},
```
Comment (AlexDBlack) [inline tests/test_ops.py:1981]: If I was doing it, to save time, I wouldn't write exhaustively every single combination of cases; I'd probably write 10-15 cases at most, that are representative (i.e., tests that vary over a bunch of different combinations shapes, ranks, datatypes, etc). Thorough is good, but there's diminishing returns eventually.
Adding empty array tests is great though. :+1:

File: tests/test_ops.py [hunk 18]
```diff
@@ -1734,9 +1790,241 @@ def test_mathtransform():
...
+        # {"opName": "bitwise_or", "outName": "bitwise_or/rank3_int32", "varShapes":[[1,1, 2], [1,1, 2]], "varTypes":["int32", "int32"], "varInit":["uniform_int10", "uniform_int2"]},
+        # {"opName": "bitwise_or", "outName": "bitwise_or/rank3_int64", "varShapes": [[1, 1, 2], [1, 1, 2]],"varTypes": ["int64", "int64"], "varInit": ["uniform_int10", "uniform_int2"]},
+
+        #{"opName": "bitwise_or", "outName": "bitwise_or/emptyArrayTest/rank2_int32", "varShapes":[[0,0], [0,0]], "varTypes":["int32", "int32"], "varInit":["uniform_int10", "empty"]},
+        #{"opName": "bitwise_or", "outName": "bitwise_or/emptyArrayTest/rank2_int64", "varShapes": [[0,0], [0,0]],"varTypes": ["int64", "int64"], "varInit": ["uniform_int10", "empty"]},
+        #{"opName": "bitwise_or", "outName": "bitwise_or/emptyArrayTest/rank3_int32", "varShapes": [[0, 0, 0], [0, 0, 0]], "varTypes": ["int32", "int32"], "varInit": ["uniform_int10", "empty"]},
+        #{"opName": "bitwise_or", "outName": "bitwise_or/emptyArrayTest/rank3_int64", "varShapes": [[0, 0, 0], [0, 0, 0]], "varTypes": ["int64", "int64"], "varInit": ["uniform_int10", "empty"]},
```
Comment (AlexDBlack) [inline tests/test_ops.py:1993]: 👍 
But switch some of these empty for empty shapes like [2, 0, 1] etc, not just the "all zeros" cases.
Same for bitwise xor and log softmax below.

File: tfoptests/ops.py [hunk 19]
```diff
@@ -1227,8 +1230,70 @@ def execute_fake_quant_with_min_max_vars_per_channel(self):
...
+
+    def execute_resize_bilinear(self):
+        return [tf.image.resize_bilinear(images = self.vars[0], size = self.vars[1])]
+
+    def execute_resize_nearest_neighbor(self):
+        return [tf.image.resize_nearest_neighbor(images = self.vars[0], size = self.vars[1])]
+
+    def execute_non_max_suppression(self):
+        return [tf.image.non_max_suppression(boxes = self.vars[0], scores = self.vars[1], max_output_size = self.vars[2],
+                                             iou_threshold=0.5, score_threshold=float('-inf'))]
```
Comment (AlexDBlack) [inline tfoptests/ops.py]: You have test cases defined with 5 inputs, but only use 3 of them here - i.e., threshold args are ignored.
Might be easier to do it for attributes with "dict" style (like your log softmax axis) for float attributes.

If we want these extra args to be optional (i.e., don't have to define them in every case) we can do something like this:
https://github.com/KonduitAI/ImportTests/blob/6c032dacc431b3ff86d41323161699d0f16497f7/tfoptests/ops.py#L749-L751

**Review by AlexDBlack — COMMENTED** [review]
The later (more recent) op tests - non-decreasing, strictly-increasing, log softmax and non_max_suppression etc - these have quite a good level of coverage, I'm happy with those :+1:
Some of the earlier ones need some work.

File: tests/test_ops.py (hunk 8 above)
Comment (alexanderst) [reply tests/test_ops.py]: Datatypes and shapes are limited here, so just added extra arguments.

File: tests/test_ops.py (hunk 9 above)
Comment (alexanderst) [reply tests/test_ops.py]: Fixed.

File: tests/test_ops.py (hunk 10 above)
Comment (alexanderst) [reply tests/test_ops.py]: Fixed.

File: tests/test_ops.py (hunk 11 above)
Comment (alexanderst) [reply tests/test_ops.py]: Fixed.

File: tests/test_ops.py (hunk 13 above)
Comment (alexanderst) [reply tests/test_ops.py]: Empty arrays tests were present before. dtype tests added.

File: tests/test_ops.py (hunk 14 above)
Comment (alexanderst) [reply tests/test_ops.py]: Added random normal inits with zero mediana.

File: tests/test_ops.py (hunk 18 above)
Comment (alexanderst) [reply tests/test_ops.py:1993]: Fixed

File: tests/test_ops.py (hunk 16 above)
Comment (alexanderst) [reply tests/test_ops.py:1602]: Added a few cases.

**Pull request merged**